### ISSUE TITLE
HTAN spike task: tile images rendering libraries review

### ISSUE THREAD

**Issue by Punzo** [OHIF/Viewers#2175]
HTAN requirements from the rendering libraries are:
1) tilesource should to be customizable, i.e.: retrieve frames (load tiles) via DICOMweb WADO-RS.
2) be able to process tile layers for applying color functions: the ideal solution would be a library which allow running custom shaders on each layer.
3) be able to blend tile layers.

While already investigated two libraries [OpenLayers](https://github.com/OHIF/Viewers/issues/2159) and [OpenSeaDragon](https://github.com/OHIF/Viewers/issues/2171), these render the data with the canvas2d [api](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage), which does not allow to customize the webGL code for point (2). Although a nonideal solution can be achive as [link](https://github.com/OHIF/Viewers/issues/2159#issuecomment-726792481), here we report an investigation of other [libraries](https://www.npmtrends.com/deck.gl-vs-leaflet-vs-openlayers-vs-mapbox-gl-vs-luma.gl), and their capabilities, checking if they can fully satisfy our requirements listed above. 

A) [deck.gl](https://deck.gl/) (see also [paper](https://osf.io/wd2gu/)): 
     .....

B) [leaflet](https://leafletjs.com/):
     .....

C) [mapbox-gl](https://docs.mapbox.com/):
     .....

**Comment by Punzo** [comment]
(B) Leaflet layers is a simple img https://github.com/Leaflet/Leaflet/blob/master/src/layer/tile/TileLayer.js#L137 or custom layers can renders the data with canvas2d api as well: https://leafletjs.com/reference-1.0.3.html

However, there are plugins:  
https://leafletjs.com/plugins.html#tile-load
https://leafletjs.com/plugins.html#tileimage-display

and for example this one, combine and apply color functions of up to 6 layers with custom shaders:
https://github.com/equinor/leaflet.tilelayer.gloperations

and then copy back the results to the canvas2d api (i.e., the same [idea](https://github.com/OHIF/Viewers/issues/2159#issuecomment-726792481) that we wanted to implement for OpenLayers): 
https://github.com/equinor/leaflet.tilelayer.gloperations/blob/338507ab827ce83f97d706e27ad39c101c295b25/src/index.ts#L2669

**Comment by Punzo** [comment]
(C) 
does not allow custom tile loader, see: https://github.com/mapbox/mapbox-gl-js/issues/2920 (requirement 1 not satisfied)

however  it is possible to define custom layer with custom shaders, see:
https://docs.mapbox.com/mapbox-gl-js/example/custom-style-layer/

**Comment by Punzo** [comment]
(A)

(i) deck.gl:

- allow custom load functions: [example1](https://developers.arcgis.com/javascript/latest/sample-code/layers-custom-tilelayer/index.html) and [example2](https://github.com/visgl/deck.gl/blob/0941bdb965e9e599dc850f1f40bce4f167ef6abb/test/apps/carto-wip/utils/carto-torque-utils.js#L49)

- allow custom layers with custom shaders: [example](https://deck.gl/docs/developer-guide/custom-layers/writing-shaders)

- good documentation 

(ii) viv library (i.e. avivator). The liabray offers loaders for Zarr and OME-TIFF data formats and a custom deck.gl Layer which applies color functions and blending for up to 6 layers.

- we still need to customize the loading for retrieving frames (load tiles) via DICOMweb WADO-RS

- [this](https://github.com/hms-dbmi/viv/blob/master/src/layers/XRLayer/XRLayer.js) is the class that perfoms the multi channel composition.

**Comment by Punzo** [comment]
additional requirements from HTAN call 20/11/2020:
1) the composition should be up to N layer
2) users should be able to give an opacity for each layer and the blending has to be take into account the opacity when adding layers.
3) loaders for Zarr and OME-TIFF will be nice to have, but it is not required.

**Comment by Punzo** [comment]
we have decided to go for the offscreen solution. Closing this

**Comment by ilan-gold** [comment]
@Punzo Hi! I am one of the maintainers of Viv.   Our docs can be a bit of a hailstorm to get through (apologies for that) and having a viewer/demo that shows the library but isn't actually part of the library as a demo site doesn't help (it masks some functionality/customizability), so sorry about that.  I just wanted to clear a few things up so at least future people who might come across this issue can get the information.  Maybe we should add a FAQ to our docs?

As a preliminary the Viv API has [multiple levels](http://viv.gehlenborglab.org/#more-in-depth) that allow for customization - you can just use the layers and then manage your deck.gl context on your own (like [vizarr](https://hms-dbmi.github.io/vizarr/?source=https://storage.googleapis.com/vitessce-demo-data/test-data/spraggins_ome.zarr) or [vitessce](http://vitessce.io/?dataset=linnarsson-2018&theme=dark)) or you can use the higher level API that Viv offers for implementing viewers (like the Picture-in-Picture or Side-by-Side, as I mention below).

>     1. number of layers for the multichannel composition limited to 6;
This number is arbitrary and can be bumped if you are interested (see https://github.com/hms-dbmi/viv/issues/304 for more info on this).  In addition, this limit is per deck.gl layer (i.e multi-channel image, with one layer per image) not across all deck.gl layers (i.e multi-channel images).  See [here](http://vitessce.io/?dataset=vanderbilt&theme=dark) for an example where we have two layers on top of one another with 8 channels total (it could be 12).  

>     2. no user provided alpha value for each layer;
[This example](http://vitessce.io/?dataset=vanderbilt&theme=dark) highlights this feature if I understand you correctly.

>     3. views limited to 2;
This is the limitation of Viv's currently implemented viewers but you can implement your own or propose a new one for Viv.  For example you could lay out 4 `SideBySideView`s (instead of 2) using the [`x` and `y` arguments](http://viv.gehlenborglab.org/#sidebysideview) with the [Views API](http://viv.gehlenborglab.org/#views), following [Viv's implementation](https://github.com/hms-dbmi/viv/blob/master/src/viewers/SideBySideViewer.js) as a guide if you want.  I'm happy to answer any questions and be of support - [here](https://portal.hubmapconsortium.org/ccf-eui) is a re-implementation of our PIP and SBS viewers in angular (our viewers are React) using our provided View classes, for example.  Alternatively, you could do something like QuPath's multi-channel viewer where you are viewing 4 channels at once from the same image, split out.  We actually have an [issue](https://github.com/hms-dbmi/viv/issues/235) for this if you are interested.

>     4. missing dicom web loaders;
Yes - if your loader implements `getTile`, `getRasterSize`, and `getRaster` like Viv's do, though, Viv's API can handle your imagery (tiled/multi-scale or not).  Perhaps this is another point of improvement, making a base abstract class that is clear in what is and isn't part of a loader.

>     5. we also we want to handle the widgets in the OIHF UI, so in multiple views we can have different combination of layers and color functions (which is not available in avivator).
I think you should be able to do this just using Viv's layers and multiple deck.gl views/canvases.  [Here](https://s3.amazonaws.com/vitessce-data/demos/2020-08-23/7280b18/index.html?url=data:,{%22name%22:%22Imaging%20Aggregation%20Methods%22,%22version%22:%221.0.0%22,%22description%22:%22Spatial%20organization%20of%20the%20somatosensory%20cortex%20revealed%20by%20cyclic%20smFISH%22,%22public%22:true,%22datasets%22:[{%22uid%22:%22vanderbilt-2020%22,%22name%22:%22Vanderbilt%202020%22,%22files%22:[{%22type%22:%22raster%22,%22fileType%22:%22raster.json%22,%22url%22:%22https://keller-mark.github.io/vitessce-demo-hosting-temporary/ome-tiff-linked-views/vanderbilt.raster.json%22}]}],%22initStrategy%22:%22auto%22,%22coordinationSpace%22:{%22spatialZoom%22:{%22A%22:-1.35956},%22spatialTargetX%22:{%22A%22:501.33924},%22spatialTargetY%22:{%22A%22:861.51522},%22spatialLayers%22:{%22ORIGINAL%22:[{%22type%22:%22raster%22,%22index%22:0,%22opacity%22:1,%22colormap%22:%22%22,%22channels%22:[{%22selection%22:{%22channel%22:0,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:1,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:2,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:3,%22z%22:0,%22time%22:0}}]}],%22GAUSSIAN%22:[{%22type%22:%22raster%22,%22index%22:1,%22opacity%22:1,%22colormap%22:%22%22,%22channels%22:[{%22selection%22:{%22channel%22:0,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:1,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:2,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:3,%22z%22:0,%22time%22:0}}]}],%22AREA%22:[{%22type%22:%22raster%22,%22index%22:2,%22opacity%22:1,%22colormap%22:%22%22,%22channels%22:[{%22selection%22:{%22channel%22:0,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:1,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:2,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:3,%22z%22:0,%22time%22:0}}]}],%22LINEAR%22:[{%22type%22:%22raster%22,%22index%22:3,%22opacity%22:1,%22colormap%22:%22%22,%22channels%22:[{%22selection%22:{%22channel%22:0,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:1,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:2,%22z%22:0,%22time%22:0}},{%22selection%22:{%22channel%22:3,%22z%22:0,%22time%22:0}}]}]}},%22layout%22:[{%22component%22:%22spatial%22,%22coordinationScopes%22:{%22spatialZoom%22:%22A%22,%22spatialTargetX%22:%22A%22,%22spatialTargetY%22:%22A%22,%22spatialLayers%22:%22ORIGINAL%22},%22x%22:0,%22y%22:0,%22w%22:3,%22h%22:4},{%22component%22:%22layerController%22,%22coordinationScopes%22:{%22spatialLayers%22:%22ORIGINAL%22},%22x%22:0,%22y%22:4,%22w%22:3,%22h%22:4},{%22component%22:%22spatial%22,%22coordinationScopes%22:{%22spatialZoom%22:%22A%22,%22spatialTargetX%22:%22A%22,%22spatialTargetY%22:%22A%22,%22spatialLayers%22:%22GAUSSIAN%22},%22x%22:3,%22y%22:0,%22w%22:3,%22h%22:4},{%22component%22:%22layerController%22,%22coordinationScopes%22:{%22spatialLayers%22:%22GAUSSIAN%22},%22x%22:3,%22y%22:4,%22w%22:3,%22h%22:4},{%22component%22:%22spatial%22,%22coordinationScopes%22:{%22spatialZoom%22:%22A%22,%22spatialTargetX%22:%22A%22,%22spatialTargetY%22:%22A%22,%22spatialLayers%22:%22AREA%22},%22x%22:6,%22y%22:0,%22w%22:3,%22h%22:4},{%22component%22:%22layerController%22,%22coordinationScopes%22:{%22spatialLayers%22:%22AREA%22},%22x%22:6,%22y%22:4,%22w%22:3,%22h%22:4},{%22component%22:%22spatial%22,%22coordinationScopes%22:{%22spatialZoom%22:%22A%22,%22spatialTargetX%22:%22A%22,%22spatialTargetY%22:%22A%22,%22spatialLayers%22:%22LINEAR%22},%22x%22:9,%22y%22:0,%22w%22:3,%22h%22:4},{%22component%22:%22layerController%22,%22coordinationScopes%22:{%22spatialLayers%22:%22LINEAR%22},%22x%22:9,%22y%22:4,%22w%22:3,%22h%22:4}]}) is an example using multiple layers acorss different deck.gl canvases, although it may take a second to load.

Lastly, because Viv exports our deck.gl layers, you can overlay things like ROI's or polygonal segmentations over your images using deck.gl `PolygonLayer`s while managing your own deck.gl context, similar to [vitessce](http://vitessce.io/?dataset=linnarsson-2018&theme=dark).  I hope this is helpful, and am excited to see whatever you all develop - offscreen rendering/ops is something I have been interested in for a while, so I'd love to see where this goes!

**Comment by pieper** [comment]
Hi @ilan-gold thanks for jumping in!  As @Punzo says we are super impressed with Viv and how deck.gl provides so many options.

We don't have a live site yet, but the offscreen work in OHIF is progressing really well in another context (PET/CT volume viewing) so we're thinking/hoping it'll be the easiest path for our microscopy work to since we already have all the dicomweb infrastructure in place.  We're implementing the offscreen stuff with vtkjs, so that gives another toybox full of filtering options and utilities to draw on.

Let's stay in touch on these things since there are a bunch of fun things left to be built and it will be great to collaborate.

**Comment by Punzo** [comment]
Hi @ilan-gold thanks for your feedback and additional information regarding the list in https://github.com/OHIF/Viewers/issues/2175#issuecomment-731278934. I completely agree viv library and deck-gl are a very promising solutions indeed! and I found your work very interesting (there are a lot of similarities with astronomy vis on which I have worked).

I should have made clear in the comment that we had two ideal candidates: the offscreen solution and the viv library. The main reasons why we picked the offscreen solution over viv are: 1) having vtkjs capabilities as Steve mentioned, 2) we want to maintain our OIHF-microscopy viewer/layout/loadDicom functionality and integrating it with the viv library is not straightforward (while the offscreen solution would be easier to integrate). So the decision was purely based on our specific needs. The project will be fully open-source, we will be very glad to share ideas and collaborate.
